third_party/libusb/webport/src/libusb_js_proxy.cc

```diff
@@ -719,109 +637,120 @@ bool CreateChromeUsbControlTransferInfo(
+    result->data_to_send = std::vector<uint8_t>(
+        transfer->buffer, transfer->buffer + transfer->length);
+  } else {
+    result->length_to_receive = transfer->length;
   }
 }
 
-chrome_usb::AsyncTransferCallback MakeChromeUsbTransferCallback(
+std::function<void(GenericRequestResult)> MakeLibusbJsTransferCallback(
     std::weak_ptr<libusb_context> context,
     const UsbTransferDestination& transfer_destination,
     LibusbJsProxy::TransferAsyncRequestStatePtr async_request_state) {
-  return
-      [context, transfer_destination, async_request_state](
-          RequestResult<chrome_usb::TransferResult> chrome_usb_request_result) {
-        const std::shared_ptr<libusb_context> locked_context = context.lock();
-        if (!locked_context) {
-          // The context that was used for the original transfer submission has
-          // been destroyed already.
-          return;
-        }
-
-        RequestResult<LibusbJsTransferResult> request_result =
-            ConvertChromeUsbTransferResultToLibusb(
-                std::move(chrome_usb_request_result));
-
-        if (transfer_destination.IsInputDirection()) {
-          locked_context->OnInputTransferResultReceived(
-              transfer_destination, std::move(request_result));
-        } else {
-          locked_context->OnOutputTransferResultReceived(
-              async_request_state, std::move(request_result));
-        }
-      };
+  return [context, transfer_destination,
+          async_request_state](GenericRequestResult js_result) {
+    const std::shared_ptr<libusb_context> locked_context = context.lock();
+    if (!locked_context) {
+      // The context that was used for the original transfer submission has
+      // been destroyed already.
+      return;
+    }
+    LibusbJsTransferResult js_transfer_result;
+    RequestResult<LibusbJsTransferResult> converted_result =
+        RemoteCallAdaptor::ConvertResultPayload(
+            std::move(js_result), &js_transfer_result, &js_transfer_result);
+    if (transfer_destination.IsInputDirection()) {
+      locked_context->OnInputTransferResultReceived(
+          transfer_destination, std::move(converted_result));
+    } else {
+      locked_context->OnOutputTransferResultReceived(
+          async_request_state, std::move(converted_result));
+    }
+  };
+}
+
+UsbTransferDestination CreateUsbTransferDestinationForTransfer(
+    libusb_transfer* transfer) {
+  GOOGLE_SMART_CARD_CHECK(transfer);
+  switch (transfer->type) {
+    case LIBUSB_TRANSFER_TYPE_CONTROL: {
+      const libusb_control_setup* const control_setup =
+          libusb_control_transfer_get_setup(transfer);
+      return UsbTransferDestination::CreateForControlTransfer(
+          transfer->dev_handle->js_device_handle(),
+          control_setup->bmRequestType, control_setup->bRequest,
+          control_setup->wValue, control_setup->wIndex);
+    }
+    case LIBUSB_TRANSFER_TYPE_BULK:
+    case LIBUSB_TRANSFER_TYPE_INTERRUPT: {
```

Comment:
nit: Not sure what C++ version we are using for SCC, but can we use [[fallthrough]] here?

Reply:
Ack - With the PNaCl compatibility we're still sitting on C++11, meanwhile `[[fallthrough]]` is a C++17 thing.

Also I think it's not a big issue here, since AFAIK this "fallthrough" is primarily intended for cases where there's some code between two case's, and hence no compiler warning.